File: PHP/quick_sort.php
<?php

/**
 * Quicksort.
 * In quick sort, we pick a random element and partition the array, such that all
 * numbers that are less than the partitioning element comes before all elements
 * that are greater than it. The partitioning can be performed efficiently through
 * a series of swaps.
 */

/**
 * Time: O(nlgn). worst case O(n^2) but less likely if random element is chosen.
 * Space: O(1)
 * 
 */
function quicksort($input) {
    // check for less than 2 elements
    if (sizeof($input) < 2) {
        return $input;
    }
    quickSortRecur($input, 0, sizeof($input)-1);
    return $input;
}

function quickSortRecur(&$input, $start, $end) {
    if ($start < $end) {
        $pivot = partition($input, $start, $end);
        quickSortRecur($input, $start, $pivot-1);
        quickSortRecur($input, $pivot+1, $end);
    }
}

function partition(&$input, $start, $end) {
    // chooses a random index as pivot
    $pivot = rand($start, $end);
    // put that element to the end
    swap($input[$pivot], $input[$end]);
    // pivot is now at the end index
    $pivot = $end;
    // last element where index value smaller than pivot
    $left = $start-1;

    for ($i=$start; $i < $end; $i++) { 
        if ($input[$i] < $input[$pivot]) {
            $left++;
            swap($input[$left], $input[$i]);
        }
    }
    swap($input[$pivot], $input[$left+1]);
    return $left+1;
} 

function swap(&$value1, &$value2) {
    $temp = $value1;
    $value1 = $value2;
    $value2 = $temp;
}

$input1 = [14, 33, 27, 10, 35, 19, 42, 44];
$input2 = [64, 25, 12, 22, 11];

echo implode(" ", quicksort($input1)) . PHP_EOL;
echo implode(" ", quicksort($input2)) . PHP_EOL;

File: Python/quicksort.py
# Partition procedure rearranges the subarray in place.


def partition(a_list, start, end):
    # last element in array
    # the element index just before the start index
    # for loop from start index to end - 1 index
    #   if current index less equal than last element
    #       i increment 1
    #       exchange less and more value positions
    # exchange more value position with the last position(pivot)
    i = start - 1
    for j in range(start, end):
        if a_list[end] > a_list[j]:
            i += 1
            a_list[j], a_list[i] = a_list[i], a_list[j]
    a_list[i+1], a_list[end] = a_list[end], a_list[i+1]
    return i + 1


def quicksort(a_list, start, end):
    if start < end:
        pivot = partition(a_list, start, end)
        quicksort(a_list, start, pivot-1)
        quicksort(a_list, pivot+1, end)
    return a_list

if __name__ == "__main__":
    test_list = [2, 8, 7, 1, 3, 5, 6, 4]
    test_list2 = [13, 19, 9, 5, 12, 8, 7, 4, 21, 2, 6, 11]
    print quicksort(test_list, 0, len(test_list)-1)
    print quicksort(test_list2, 0, len(test_list2)-1)
